File: PathoFroh/src/main/java/com/patho/main/action/dialog/slides/StainingPhaseExitDialog.java

package com.patho.main.action.dialog.slides;

import com.patho.main.action.dialog.AbstractDialog;
import com.patho.main.common.Dialog;
import com.patho.main.common.PredefinedFavouriteList;
import com.patho.main.model.patient.Task;
import com.patho.main.service.FavouriteListService;
import com.patho.main.util.event.dialog.StainingPhaseExitEvent;
import com.patho.main.util.task.TaskStatus;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.Setter;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Configurable;

@Configurable
@Getter
@Setter
public class StainingPhaseExitDialog extends AbstractDialog {

    @Autowired
    @Getter(AccessLevel.NONE)
    @Setter(AccessLevel.NONE)
    private FavouriteListService favouriteListService;

    /**
     * If true the task will be removed from worklist
     */
    private boolean removeFromWorklist;

    /**
     * If true the task will be removed from staining list
     */
    private boolean removeFromStainingList;

    /**
     * If true the staining phase of the task will be finished
     */
    private boolean endStainingPhase;

    /**
     * If true the task will be shifted to the diagnosis phase
     */
    private boolean goToDiagnosisPhase;

    /**
     * Initializes the bean and shows the dialog
     */
    public StainingPhaseExitDialog initAndPrepareBean(Task task) {
        return initAndPrepareBean(task, false);
    }

    public StainingPhaseExitDialog initAndPrepareBean(Task task, boolean autoCompleteStainings) {
        if (initBean(task, autoCompleteStainings))
            prepareDialog();
        return this;
    }

    /**
     * Initializes all field of the object
     */
    public boolean initBean(Task task, boolean autoCompleteStainings) {

        if (TaskStatus.hasFavouriteLists(task, PredefinedFavouriteList.NotificationList))
            goToDiagnosisPhase = false;
        else
            goToDiagnosisPhase = true;

        // all slides will be marked as completed by endStainingphase methode
        if (autoCompleteStainings) {
            setRemoveFromStainingList(true);
            setRemoveFromWorklist(true);
            setEndStainingPhase(true);
        } else {
            boolean stainingCompleted = TaskStatus.checkIfStainingCompleted(task);

            setRemoveFromStainingList(stainingCompleted);
            setRemoveFromWorklist(stainingCompleted);
            setEndStainingPhase(stainingCompleted);
        }

        //data.setGoToDiagnosisPhase(true);

        return super.initBean(task, Dialog.STAINING_PHASE_EXIT);
    }

    public void exitPhaseAndClose() {
        super.hideDialog(new StainingPhaseExitEvent(task, removeFromWorklist, removeFromStainingList, endStainingPhase, goToDiagnosisPhase));
    }

}

File: src/main/kotlin/com/patho/main/dialog/phase/StainingPhaseExitDialog.kt
package com.patho.main.dialog.phase

import com.patho.main.action.handler.WorkPhaseHandler
import com.patho.main.common.Dialog
import com.patho.main.model.patient.Task
import com.patho.main.repository.jpa.TaskRepository
import com.patho.main.service.SlideService
import com.patho.main.util.dialog.event.StainingPhaseExitEvent
import com.patho.main.util.dialog.event.TaskReloadEvent
import com.patho.main.util.status.TotalTaskStatus
import com.patho.main.util.ui.backend.CheckBoxStatus
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Component

/**
 * Dialog for leaving the staining phase
 */
@Component()
@Scope(value = "session")
open class StainingPhaseExitDialog @Autowired constructor(
        private val taskRepository: TaskRepository,
        private val workPhaseHandler: WorkPhaseHandler,
        private val slideService: SlideService) : AbstractPhaseExitDialog(Dialog.STAINING_PHASE_EXIT) {

    /**
     * Checkbox backend for completeStainings
     */
    open lateinit var completeStainings: CheckBoxStatus

    /**
     * Checkbox backend for goToDiagnosis
     */
    open lateinit var goToDiagnosis: CheckBoxStatus

    /**
     * Initializes the dialog
     */
    override fun initBean(task: Task): Boolean {
        val oTask = taskRepository.findByID(task.id, true, true, false, true, true)

        val status = TotalTaskStatus(oTask)

        // complete checkbox
        completeStainings = object : CheckBoxStatus() {
            override fun onClick() {
                exitPhase.set(completeStainings.value, true, !completeStainings.value, false)
                goToDiagnosis.set(true, true, false, false)
                removeFromWorklist.set(completeStainings.value, true, false, false)
            }
        }

        // diagnosis checkbox
        goToDiagnosis = object : CheckBoxStatus() {
            override fun onClick() {
                isInfo = exitPhase.value && !goToDiagnosis.value
            }
        }

        // complete stainings if not all completed
        completeStainings.set(true, !status.staining.slidesMarkedAsCompleted, false, false)

        // remove from list
        removeFromWorklist.set(true, true, false, false)

        // goto diagnosis pahse
        goToDiagnosis.set(true, true, false, false)

        // exit phase
        exitPhase.set(completeStainings.value, true, !completeStainings.value, false)

        return super.initBean(oTask)
    }

    /**
     * Hides the dialog and removes the task from the staining lsit
     */
    open fun hideAndExitPhase() {
        var task = if (completeStainings.value) slideService.completeStaining(task, true, true) else task
        task = workPhaseHandler.endStainingPhase(task, exitPhase.value, goToDiagnosis.value, removeFromWorklist.value)
        super.hideDialog(StainingPhaseExitEvent())
    }

    /**
     * Hides the dialog, fires a task reloadTaskEvent
     */
    override fun hideDialog() {
        super.hideDialog(TaskReloadEvent())
    }
}
